components/tidb_query/src/rpn_expr/impl_json.rs

```diff
@@ -17,6 +17,67 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
         .map(|json_arg| Bytes::from(json_arg.json_type())))
 }
 
+#[rpn_fn(raw_varg, min_args = 2, extra_validator = json_modify_validator)]
+#[inline]
+fn json_set(args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    json_modify(args, ModifyType::Set)
+}
+
+#[rpn_fn(raw_varg, min_args = 2, extra_validator = json_modify_validator)]
+#[inline]
+fn json_insert(args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    json_modify(args, ModifyType::Insert)
+}
+
+#[rpn_fn(raw_varg, min_args = 2, extra_validator = json_modify_validator)]
+#[inline]
+fn json_replace(args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    json_modify(args, ModifyType::Replace)
+}
+
+#[inline]
+fn json_modify(args: &[ScalarValueRef], mt: ModifyType) -> Result<Option<Json>> {
+    assert!(args.len() >= 2);
+    // base Json argument
+    let base: &Option<Json> = args[0].as_ref();
+    let mut base = base.as_ref().map_or(Json::None, |json| json.to_owned());
+
+    let buf_size = args.len() / 2;
+
+    let mut path_expr_list = Vec::with_capacity(buf_size);
+    let mut values = Vec::with_capacity(buf_size);
+
+    for chunk in args[1..].chunks(2) {
+        let path: &Option<Bytes> = chunk[0].as_ref();
+        let value: &Option<Json> = chunk[1].as_ref();
+
+        path_expr_list.push(try_opt!(parse_json_path(path)));
+
+        let value = value.as_ref().map_or(Json::None, |json| json.to_owned());
+        values.push(value);
+    }
+    base.modify(&path_expr_list, values, mt)?;
+
+    Ok(Some(base))
+}
+
+/// validate the arguments are `(&Option<Json>, &[(Option<Bytes>, Option<Json>)])`
+fn json_modify_validator(expr: &tipb::Expr) -> Result<()> {
+    let children = expr.get_children();
+    assert!(children.len() >= 2);
+    if children.len() % 2 != 1 {
```

Comment:
Could we apply a pattern like: `if x & 1 == 0`

Reply:
> Could we apply a pattern like: `if x & 1 == 0`?

I think using it may be confusing. And compiler will help us optimize `children.len() % 2 != 1`.

Reply:
> Could we apply a pattern like: `if x & 1 == 0`?

I found that in GCC, it will optimize it: https://stackoverflow.com/questions/22446425/do-c-c-compilers-such-as-gcc-generally-optimize-modulo-by-a-constant-power-of